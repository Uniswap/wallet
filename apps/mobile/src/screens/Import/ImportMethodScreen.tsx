import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { TFunction } from 'i18next'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { Alert } from 'react-native'
import { useAppDispatch, useAppTheme } from 'src/app/hooks'
import { OnboardingStackParamList } from 'src/app/navigation/types'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import Trace from 'src/components/Trace/Trace'
import { isICloudAvailable } from 'src/features/CloudBackup/RNICloudBackupsManager'
import { OnboardingScreen } from 'src/features/onboarding/OnboardingScreen'
import { OptionCard } from 'src/features/onboarding/OptionCard'
import { ImportType, OnboardingEntryPoint } from 'src/features/onboarding/utils'
import { ElementName } from 'src/features/telemetry/constants'
import { OnboardingScreens } from 'src/screens/Screens'
import { openSettings } from 'src/utils/linking'
import { useAddBackButton } from 'src/utils/useAddBackButton'
import EyeIcon from 'ui/src/assets/icons/eye.svg'
import ImportIcon from 'ui/src/assets/icons/paper-stack.svg'
import { Theme } from 'ui/src/theme/restyle/theme'
import {
  PendingAccountActions,
  pendingAccountActions,
} from 'wallet/src/features/wallet/create/pendingAccountsSaga'

interface ImportMethodOption {
  title: (t: TFunction) => string
  blurb: (t: TFunction) => string
  icon: (theme: Theme) => React.ReactNode
  nav: OnboardingScreens
  importType: ImportType
  name: ElementName
  badgeText?: (t: TFunction) => string
}

const options: ImportMethodOption[] = [
  {
    title: (t: TFunction) => t('Import a wallet'),
    blurb: (t: TFunction) => t('Enter your recovery phrase from another crypto wallet'),
    icon: (theme: Theme) => (
      <ImportIcon color={theme.colors.magentaVibrant} height={18} strokeWidth="1.5" width={18} />
    ),
    nav: OnboardingScreens.SeedPhraseInput,
    importType: ImportType.SeedPhrase,
    name: ElementName.OnboardingImportSeedPhrase,
    badgeText: (t: TFunction) => t('Recommended'),
  },
  {
    title: (t: TFunction) => t('Watch a wallet'),
    blurb: (t: TFunction) =>
      t('Explore the contents of a wallet by entering any address or ENS name '),
    icon: (theme: Theme) => (
      <EyeIcon color={theme.colors.magentaVibrant} height={24} strokeWidth="1.5" width={24} />
    ),
    nav: OnboardingScreens.WatchWallet,
    importType: ImportType.Watch,
    name: ElementName.OnboardingImportWatchedAccount,
  },
]

type Props = NativeStackScreenProps<OnboardingStackParamList, OnboardingScreens.ImportMethod>

export function ImportMethodScreen({ navigation, route: { params } }: Props): JSX.Element {
  const { t } = useTranslation()
  const theme = useAppTheme()
  const dispatch = useAppDispatch()
  const entryPoint = params?.entryPoint

  useAddBackButton(navigation)

  const handleOnPressRestoreBackup = async (): Promise<void> => {
    const iCloudAvailable = await isICloudAvailable()

    if (!iCloudAvailable) {
      Alert.alert(
        t('iCloud Drive not available'),
        t(
          'Please verify that you are logged in to an Apple ID with iCloud Drive enabled on this device and try again.'
        ),
        [
          { text: t('Go to settings'), onPress: openSettings, style: 'default' },
          { text: t('Not now'), style: 'cancel' },
        ]
      )
      return
    }

    navigation.navigate({
      name: OnboardingScreens.RestoreCloudBackupLoading,
      params: { importType: ImportType.Restore, entryPoint },
      merge: true,
    })
  }

  const handleOnPress = async (nav: OnboardingScreens, importType: ImportType): Promise<void> => {
    // Delete any pending accounts before entering flow.
    dispatch(pendingAccountActions.trigger(PendingAccountActions.Delete))

    if (importType === ImportType.Restore) {
      await handleOnPressRestoreBackup()
      return
    }

    navigation.navigate({
      name: nav,
      params: { importType, entryPoint },
      merge: true,
    })
  }

  const importOptions =
    entryPoint === OnboardingEntryPoint.Sidebar
      ? options.filter((option) => option.name !== ElementName.OnboardingImportWatchedAccount)
      : options

  return (
    <OnboardingScreen title={t('How do you want to add your wallet?')}>
      <Flex grow gap="spacing12" marginTop="spacing4">
        {importOptions.map(({ title, blurb, icon, nav, importType, name, badgeText }) => (
          <OptionCard
            key={'connection-option-' + title}
            hapticFeedback
            badgeText={badgeText?.(t)}
            blurb={blurb(t)}
            elementName={name}
            icon={icon(theme)}
            title={title(t)}
            onPress={(): Promise<void> => handleOnPress(nav, importType)}
          />
        ))}
      </Flex>
      <Trace logPress element={ElementName.OnboardingImportBackup}>
        <TouchableArea alignItems="center" mb="spacing12">
          <Text
            color="accentAction"
            variant="buttonLabelMedium"
            onPress={(): Promise<void> =>
              handleOnPress(OnboardingScreens.RestoreCloudBackup, ImportType.Restore)
            }>
            {t('Restore from iCloud')}
          </Text>
        </TouchableArea>
      </Trace>
    </OnboardingScreen>
  )
}
