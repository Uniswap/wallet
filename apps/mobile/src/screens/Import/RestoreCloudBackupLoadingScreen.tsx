import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React, { useCallback, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useAppTheme } from 'src/app/hooks'
import { OnboardingStackParamList } from 'src/app/navigation/types'
import { Box } from 'src/components/layout'
import { BaseCard } from 'src/components/layout/BaseCard'
import { Loader } from 'src/components/loading'
import { useCloudBackups } from 'src/features/CloudBackup/hooks'
import {
  startFetchingICloudBackups,
  stopFetchingICloudBackups,
} from 'src/features/CloudBackup/RNICloudBackupsManager'
import { OnboardingScreen } from 'src/features/onboarding/OnboardingScreen'
import { ImportType } from 'src/features/onboarding/utils'
import { OnboardingScreens } from 'src/screens/Screens'
import { useAddBackButton } from 'src/utils/useAddBackButton'
import CloudIcon from 'ui/src/assets/icons/cloud.svg'
import { logger } from 'wallet/src/features/logger/logger'
import { useAsyncData } from 'wallet/src/utils/hooks'
import { ONE_SECOND_MS } from 'wallet/src/utils/time'
import { useTimeout } from 'wallet/src/utils/timing'

type Props = NativeStackScreenProps<
  OnboardingStackParamList,
  OnboardingScreens.RestoreCloudBackupLoading
>

const MIN_LOADING_UI_MS = ONE_SECOND_MS
// 10s timeout time for query for backups, since we don't know when the query completes
const MAX_LOADING_TIMEOUT_MS = ONE_SECOND_MS * 10

export function RestoreCloudBackupLoadingScreen({
  navigation,
  route: { params },
}: Props): JSX.Element {
  const { t } = useTranslation()
  const theme = useAppTheme()
  const entryPoint = params.entryPoint

  const [isLoading, setIsLoading] = useState(true)
  const backups = useCloudBackups()

  useAddBackButton(navigation)

  // Starts query for iCloud backup files, backup files found are streamed into Redux
  const fetchICloudBackupsWithTimeout = useCallback(async () => {
    // Show loading state for max 10s, then show no backups found
    setIsLoading(true)
    await startFetchingICloudBackups()

    setTimeout(async () => {
      logger.debug(
        'RestoreCloudBackupLoadingScreen',
        'fetchICloudBackupsWithTimeout',
        `Timed out fetching iCloud backups after ${MAX_LOADING_TIMEOUT_MS}ms`
      )
      setIsLoading(false)
      await stopFetchingICloudBackups()
    }, MAX_LOADING_TIMEOUT_MS)
  }, [])

  useAsyncData(fetchICloudBackupsWithTimeout)
  // After finding backups, show loading state for minimum 1s to prevent screen changing too quickly
  useTimeout(
    backups.length > 0
      ? (): void => {
          if (backups.length === 1 && backups[0]) {
            navigation.replace(OnboardingScreens.RestoreCloudBackupPassword, {
              importType: ImportType.Restore,
              entryPoint,
              mnemonicId: backups[0].mnemonicId,
            })
          } else {
            navigation.replace(OnboardingScreens.RestoreCloudBackup, {
              importType: ImportType.Restore,
              entryPoint,
            })
          }
        }
      : (): undefined => undefined,
    MIN_LOADING_UI_MS
  )

  // Handle no backups found error state
  if (!isLoading && backups.length === 0) {
    return (
      <Box alignSelf="center" px="spacing16">
        <BaseCard.ErrorState
          description={t(`It looks like you haven't backed up any of your seed phrases to iCloud.`)}
          icon={
            <CloudIcon
              color={theme.colors.textTertiary}
              height={theme.imageSizes.image48}
              width={theme.imageSizes.image48}
            />
          }
          retryButtonLabel={t('Retry')}
          title={t('0 backups found')}
          onRetry={fetchICloudBackupsWithTimeout}
        />
      </Box>
    )
  }

  return (
    <OnboardingScreen title={t('Searching for backups...')}>
      <Loader.Wallets repeat={5} />
    </OnboardingScreen>
  )
}
