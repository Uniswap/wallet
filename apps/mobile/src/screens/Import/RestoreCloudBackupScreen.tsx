import { NativeStackScreenProps } from '@react-navigation/native-stack'
import dayjs from 'dayjs'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { ScrollView } from 'react-native-gesture-handler'
import { useAppDispatch, useAppTheme } from 'src/app/hooks'
import { OnboardingStackParamList } from 'src/app/navigation/types'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Chevron } from 'src/components/icons/Chevron'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { Unicon } from 'src/components/unicons/Unicon'
import { useCloudBackups } from 'src/features/CloudBackup/hooks'
import { ICloudMnemonicBackup } from 'src/features/CloudBackup/types'
import { OnboardingScreen } from 'src/features/onboarding/OnboardingScreen'
import { OnboardingScreens } from 'src/screens/Screens'
import { useAddBackButton } from 'src/utils/useAddBackButton'
import {
  PendingAccountActions,
  pendingAccountActions,
} from 'wallet/src/features/wallet/create/pendingAccountsSaga'
import { shortenAddress } from 'wallet/src/utils/addresses'

type Props = NativeStackScreenProps<OnboardingStackParamList, OnboardingScreens.RestoreCloudBackup>

export function RestoreCloudBackupScreen({ navigation, route: { params } }: Props): JSX.Element {
  const { t } = useTranslation()
  const theme = useAppTheme()
  const dispatch = useAppDispatch()
  const backups = useCloudBackups()
  const sortedBackups = backups.slice().sort((a, b) => b.createdAt - a.createdAt)

  const onPressRestoreBackup = async (backup: ICloudMnemonicBackup): Promise<void> => {
    // Clear any existing pending accounts
    dispatch(pendingAccountActions.trigger(PendingAccountActions.Delete))

    navigation.navigate({
      name: OnboardingScreens.RestoreCloudBackupPassword,
      params: { ...params, mnemonicId: backup.mnemonicId },
      merge: true,
    })
  }

  useAddBackButton(navigation)

  return (
    <OnboardingScreen
      subtitle={t('There are multiple recovery phrases backed up to your iCloud.')}
      title={t('Select backup to restore')}>
      <ScrollView>
        <Flex gap="spacing8">
          {sortedBackups.map((backup, index) => {
            const { mnemonicId, createdAt } = backup
            return (
              <TouchableArea
                key={backup.mnemonicId}
                backgroundColor="background2"
                borderColor="background3"
                borderRadius="rounded16"
                borderWidth={1}
                p="spacing16"
                onPress={(): Promise<void> => onPressRestoreBackup(backup)}>
                <Flex row alignItems="center" justifyContent="space-between">
                  <Flex centered row gap="spacing12">
                    <Unicon address={mnemonicId} size={32} />
                    <Flex gap="none">
                      <Text numberOfLines={1} variant="subheadSmall">
                        {t('Backup {{backupIndex}}', { backupIndex: sortedBackups.length - index })}
                      </Text>
                      <Text color="textSecondary" variant="buttonLabelMicro">
                        {shortenAddress(mnemonicId)}
                      </Text>
                    </Flex>
                  </Flex>
                  <Flex row gap="spacing12">
                    <Flex alignItems="flex-end" gap="spacing4">
                      <Text color="textSecondary" variant="buttonLabelMicro">
                        {t('Backed up on:')}
                      </Text>
                      <Text variant="buttonLabelMicro">
                        {dayjs.unix(createdAt).format('MMM D, YYYY, h:mma')}
                      </Text>
                    </Flex>
                    <Chevron color={theme.colors.textPrimary} direction="e" />
                  </Flex>
                </Flex>
              </TouchableArea>
            )
          })}
        </Flex>
      </ScrollView>
    </OnboardingScreen>
  )
}
