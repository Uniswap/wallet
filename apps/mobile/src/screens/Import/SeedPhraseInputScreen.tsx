import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { useResponsiveProp } from '@shopify/restyle'
import React, { useCallback, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useAppDispatch } from 'src/app/hooks'
import { OnboardingStackParamList } from 'src/app/navigation/types'
import { Button } from 'src/components/buttons/Button'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import Trace from 'src/components/Trace/Trace'
import { RECOVERY_PHRASE_HELP_URL } from 'src/constants/urls'
import { useLockScreenOnBlur } from 'src/features/authentication/lockScreenContext'
import { GenericImportForm } from 'src/features/import/GenericImportForm'
import { SafeKeyboardOnboardingScreen } from 'src/features/onboarding/SafeKeyboardOnboardingScreen'
import { ElementName } from 'src/features/telemetry/constants'
import { OnboardingScreens } from 'src/screens/Screens'
import { openUri } from 'src/utils/linking'
import { useAddBackButton } from 'src/utils/useAddBackButton'
import { importAccountActions } from 'wallet/src/features/wallet/import/importAccountSaga'
import { ImportAccountType } from 'wallet/src/features/wallet/import/types'
import { NUMBER_OF_WALLETS_TO_IMPORT } from 'wallet/src/features/wallet/import/utils'
import {
  MnemonicValidationError,
  translateMnemonicErrorMessage,
  userFinishedTypingWord,
  validateMnemonic,
  validateSetOfWords,
} from 'wallet/src/utils/mnemonics'

type Props = NativeStackScreenProps<OnboardingStackParamList, OnboardingScreens.ImportMethod>

export function SeedPhraseInputScreen({ navigation, route: { params } }: Props): JSX.Element {
  const dispatch = useAppDispatch()
  const { t } = useTranslation()

  /**
   * If paste permission modal is open, we need to manually disable the splash screen that appears on blur,
   * since the modal triggers the same `inactive` app state as does going to app switcher
   *
   * Technically seed phrase will be blocked if user pastes from keyboard,
   * but that is an extreme edge case.
   **/
  const [pastePermissionModalOpen, setPastePermissionModalOpen] = useState(false)
  useLockScreenOnBlur(pastePermissionModalOpen)

  const [value, setValue] = useState<string | undefined>(undefined)
  const [showSuccess, setShowSuccess] = useState(false)
  const [errorMessage, setErrorMessage] = useState<string | undefined>(undefined)

  useAddBackButton(navigation)

  // Add all accounts from mnemonic.
  const onSubmit = useCallback(() => {
    // Check phrase validation
    const { validMnemonic, error, invalidWord } = validateMnemonic(value)

    if (error) {
      setErrorMessage(translateMnemonicErrorMessage(error, invalidWord, t))
      return
    }

    dispatch(
      importAccountActions.trigger({
        type: ImportAccountType.Mnemonic,
        validatedMnemonic: validMnemonic,
        indexes: Array.from(Array(NUMBER_OF_WALLETS_TO_IMPORT).keys()),
      })
    )
    navigation.navigate({ name: OnboardingScreens.SelectWallet, params, merge: true })
  }, [value, t, dispatch, navigation, params])

  const onBlur = useCallback(() => {
    const { error, invalidWord } = validateMnemonic(value)
    if (error) {
      setShowSuccess(false)
      setErrorMessage(translateMnemonicErrorMessage(error, invalidWord, t))
    }
  }, [t, value])

  const onChange = (text: string | undefined): void => {
    const { error, invalidWord, isValidLength } = validateSetOfWords(text)

    // always show success UI if phrase is valid length
    if (isValidLength) {
      setShowSuccess(true)
    } else {
      setShowSuccess(false)
    }

    // suppress error messages if the  user is not done typing a word
    const suppressError =
      (error === MnemonicValidationError.InvalidWord && !userFinishedTypingWord(text)) ||
      error === MnemonicValidationError.NotEnoughWords

    if (!error || suppressError) {
      setErrorMessage(undefined)
    } else {
      setErrorMessage(translateMnemonicErrorMessage(error, invalidWord, t))
    }

    setValue(text)
  }

  const onPressRecoveryHelpButton = (): Promise<void> => openUri(RECOVERY_PHRASE_HELP_URL)

  const subtitleSize = useResponsiveProp({
    xs: 'bodyMicro',
    sm: 'subheadSmall',
  })

  const itemSpacing = useResponsiveProp({
    xs: 'none',
    sm: 'spacing8',
  })

  return (
    <SafeKeyboardOnboardingScreen
      subtitle={t('Your recovery phrase will only be stored locally on your device.')}
      title={t('Enter your recovery phrase')}>
      <Flex gap={itemSpacing}>
        <GenericImportForm
          autoCorrect
          blurOnSubmit
          liveCheck
          afterPasteButtonPress={(): void => setPastePermissionModalOpen(false)}
          beforePasteButtonPress={(): void => setPastePermissionModalOpen(true)}
          errorMessage={errorMessage}
          placeholderLabel={t('recovery phrase')}
          showSuccess={showSuccess}
          textAlign="center" // Fixes iOS text input issue where right align trims trailing whitespace
          value={value}
          onBlur={onBlur}
          onChange={onChange}
        />
        <Flex centered>
          <Trace logPress element={ElementName.RecoveryHelpButton}>
            <TouchableArea onPress={onPressRecoveryHelpButton}>
              <Text color="accentAction" variant={subtitleSize}>
                {t('How do I find my recovery phrase?')}
              </Text>
            </TouchableArea>
          </Trace>
        </Flex>
      </Flex>

      <Trace logPress element={ElementName.Next}>
        <Button disabled={!!errorMessage || !value} label={t('Continue')} onPress={onSubmit} />
      </Trace>
    </SafeKeyboardOnboardingScreen>
  )
}
