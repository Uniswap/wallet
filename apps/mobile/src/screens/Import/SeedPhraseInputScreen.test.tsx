import { CompositeNavigationProp, RouteProp } from '@react-navigation/core'
import { NativeStackNavigationProp } from '@react-navigation/native-stack'
import { StackNavigationProp } from '@react-navigation/stack'
import React from 'react'
import { AppStackParamList, OnboardingStackParamList } from 'src/app/navigation/types'
import { SeedPhraseInputScreen } from 'src/screens/Import/SeedPhraseInputScreen'
import { OnboardingScreens, Screens } from 'src/screens/Screens'
import { render } from 'src/test/test-utils'

jest.mock('src/utils/useAddBackButton', () => ({
  useAddBackButton: (): jest.Mock => jest.fn(),
}))

const navigationProp = {} as CompositeNavigationProp<
  StackNavigationProp<OnboardingStackParamList, OnboardingScreens.ImportMethod, undefined>,
  NativeStackNavigationProp<AppStackParamList, Screens.OnboardingStack, undefined>
>

const routeProp = {} as RouteProp<OnboardingStackParamList, OnboardingScreens.ImportMethod>

describe(SeedPhraseInputScreen, () => {
  it('seed phrase initial screen rendering', async () => {
    const tree = render(<SeedPhraseInputScreen navigation={navigationProp} route={routeProp} />)

    expect(tree.toJSON()).toMatchSnapshot()
  })
})
