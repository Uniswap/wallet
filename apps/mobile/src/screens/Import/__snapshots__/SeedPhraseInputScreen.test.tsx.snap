// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeedPhraseInputScreen seed phrase initial screen rendering 1`] = `
<View
  sentry-label="Screen"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "display": "flex",
        "flex": 1,
      },
      {
        "paddingLeft": undefined,
        "paddingRight": undefined,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          {
            "flex": 1,
            "justifyContent": "flex-end",
          },
          {
            "marginBottom": 10,
          },
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      bounces={false}
      contentContainerStyle={
        {
          "flexGrow": 1,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <View
          entering={BaseAnimationMock {}}
          exiting={BaseAnimationMock {}}
          sentry-label="Flex"
          style={
            [
              {
                "alignItems": "stretch",
                "display": "flex",
                "flexDirection": "column",
                "justifyContent": "flex-start",
                "minHeight": 0,
                "paddingBottom": 16,
                "paddingHorizontal": 16,
              },
              [
                {
                  "flexGrow": 1,
                },
                {
                  "paddingTop": 200,
                },
              ],
            ]
          }
        >
          <View
            sentry-label="Flex"
            style={
              [
                {
                  "alignItems": "stretch",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                  "margin": 12,
                },
              ]
            }
          >
            <Text
              allowFontScaling={true}
              maxFontSizeMultiplier={1.2}
              style={
                [
                  {
                    "color": "#0E1524",
                    "display": "flex",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 24,
                    "lineHeight": 28,
                    "paddingTop": 0,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Enter your recovery phrase
            </Text>
            <View
              style={
                [
                  {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 12,
                  },
                ]
              }
            />
            <Text
              allowFontScaling={true}
              maxFontSizeMultiplier={1.4}
              style={
                [
                  {
                    "color": "#5D6785",
                    "display": "flex",
                    "fontFamily": "Inter-Regular",
                    "fontSize": 15,
                    "lineHeight": 20,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Your recovery phrase will only be stored locally on your device.
            </Text>
          </View>
          <View
            style={
              [
                {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 16,
                },
              ]
            }
          />
          <View
            sentry-label="Flex"
            style={
              [
                {
                  "alignItems": "stretch",
                  "display": "flex",
                  "flexDirection": "column",
                  "flexGrow": 1,
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              sentry-label="Flex"
              style={
                [
                  {
                    "alignItems": "stretch",
                    "display": "flex",
                    "flexDirection": "column",
                    "justifyContent": "flex-start",
                  },
                ]
              }
            >
              <View
                onTouchEnd={[Function]}
                sentry-label="Flex"
                style={
                  [
                    {
                      "alignItems": "stretch",
                      "display": "flex",
                      "flexDirection": "column",
                      "justifyContent": "flex-start",
                    },
                  ]
                }
              >
                <View
                  sentry-label="Flex"
                  style={
                    [
                      {
                        "alignItems": "center",
                        "backgroundColor": "#FFFFFF",
                        "borderColor": "#F5F6FC",
                        "borderRadius": 20,
                        "borderWidth": 1,
                        "display": "flex",
                        "flexDirection": "column",
                        "flexShrink": 1,
                        "justifyContent": "center",
                        "minHeight": 120,
                        "paddingHorizontal": 36,
                        "paddingVertical": 16,
                        "width": "100%",
                      },
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "flex-end",
                          "display": "flex",
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "width": "100%",
                        },
                      ]
                    }
                  >
                    <TextInput
                      autoCapitalize="none"
                      autoComplete="off"
                      autoCorrect={true}
                      autoFocus={true}
                      blurOnSubmit={true}
                      maxFontSizeMultiplier={1.4}
                      multiline={true}
                      onBlur={[Function]}
                      onChangeText={[Function]}
                      onFocus={[Function]}
                      onSubmitEditing={[Function]}
                      placeholderTextColor="#888FAB"
                      returnKeyType="done"
                      scrollEnabled={false}
                      selectionColor="#0E1524"
                      spellCheck={false}
                      style={
                        [
                          {
                            "backgroundColor": "transparent",
                            "borderRadius": 12,
                            "color": "#0E1524",
                            "fontSize": 17,
                            "justifyContent": "center",
                            "lineHeight": 17,
                            "paddingHorizontal": 0,
                            "paddingVertical": 0,
                            "textAlign": "center",
                            "width": 8,
                          },
                        ]
                      }
                      testID="import_account_form/input"
                      textAlignVertical="bottom"
                    />
                  </View>
                  <View
                    style={
                      [
                        {
                          "display": "flex",
                          "flexGrow": 0,
                          "flexShrink": 0,
                          "height": 0,
                        },
                      ]
                    }
                  />
                  <View
                    onLayout={[Function]}
                    sentry-label="Flex"
                    style={
                      [
                        {
                          "alignItems": "center",
                          "display": "flex",
                          "flexDirection": "row",
                          "flexGrow": 1,
                          "justifyContent": "center",
                          "paddingTop": 4,
                          "position": "absolute",
                        },
                      ]
                    }
                  >
                    <Text
                      adjustsFontSizeToFit={true}
                      allowFontScaling={true}
                      maxFontSizeMultiplier={1.4}
                      numberOfLines={1}
                      style={
                        [
                          {
                            "color": "#5D6785",
                            "display": "flex",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 17,
                            "lineHeight": 24,
                          },
                          {
                            "flexShrink": 1,
                          },
                        ]
                      }
                    >
                      Type or
                    </Text>
                    <View
                      style={
                        [
                          {
                            "display": "flex",
                            "flexGrow": 0,
                            "flexShrink": 0,
                            "height": 8,
                            "width": 8,
                          },
                        ]
                      }
                    />
                    <View
                      accessibilityState={
                        {
                          "disabled": false,
                        }
                      }
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      hitSlop={
                        {
                          "bottom": 5,
                          "left": 5,
                          "right": 5,
                          "top": 5,
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        {
                          "alignItems": "center",
                          "backgroundColor": "#E8ECFB",
                          "borderColor": "transparent",
                          "borderRadius": 8,
                          "borderWidth": 1,
                          "display": "flex",
                          "opacity": 1,
                        }
                      }
                    >
                      <View
                        sentry-label="Flex"
                        style={
                          [
                            {
                              "alignItems": "stretch",
                              "display": "flex",
                              "flexDirection": "row",
                              "justifyContent": "flex-start",
                              "paddingHorizontal": 8,
                              "paddingVertical": 8,
                            },
                          ]
                        }
                      >
                        <svg
                          color="#0E1524"
                          data-file-name="SvgPaste"
                          height={16}
                          strokeWidth={2}
                          width={16}
                        />
                        <View
                          style={
                            [
                              {
                                "display": "flex",
                                "flexGrow": 0,
                                "flexShrink": 0,
                                "height": 4,
                                "width": 4,
                              },
                            ]
                          }
                        />
                        <Text
                          allowFontScaling={true}
                          maxFontSizeMultiplier={1.2}
                          style={
                            [
                              {
                                "color": "#0E1524",
                                "display": "flex",
                                "fontFamily": "Inter-SemiBold",
                                "fontSize": 15,
                                "lineHeight": 20,
                                "textAlign": "center",
                              },
                            ]
                          }
                        >
                          Paste
                        </Text>
                      </View>
                    </View>
                    <View
                      style={
                        [
                          {
                            "display": "flex",
                            "flexGrow": 0,
                            "flexShrink": 0,
                            "height": 8,
                            "width": 8,
                          },
                        ]
                      }
                    />
                    <Text
                      adjustsFontSizeToFit={true}
                      allowFontScaling={true}
                      maxFontSizeMultiplier={1.4}
                      numberOfLines={1}
                      style={
                        [
                          {
                            "color": "#5D6785",
                            "display": "flex",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 17,
                            "lineHeight": 24,
                          },
                          {
                            "flexShrink": 1,
                          },
                        ]
                      }
                    >
                      recovery phrase
                    </Text>
                  </View>
                </View>
                <View
                  style={
                    [
                      {
                        "display": "flex",
                        "flexGrow": 0,
                        "flexShrink": 0,
                        "height": 16,
                      },
                    ]
                  }
                />
                <View
                  sentry-label="Flex"
                  style={
                    [
                      {
                        "alignItems": "stretch",
                        "display": "flex",
                        "flexDirection": "column",
                        "justifyContent": "flex-start",
                      },
                    ]
                  }
                />
              </View>
              <View
                style={
                  [
                    {
                      "display": "flex",
                      "flexGrow": 0,
                      "flexShrink": 0,
                      "height": 8,
                    },
                  ]
                }
              />
              <View
                sentry-label="Flex"
                style={
                  [
                    {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "column",
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  hitSlop={
                    {
                      "bottom": 5,
                      "left": 5,
                      "right": 5,
                      "top": 5,
                    }
                  }
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "display": "flex",
                      "opacity": 1,
                    }
                  }
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    style={
                      [
                        {
                          "color": "#FC72FF",
                          "display": "flex",
                          "fontFamily": "Inter-Medium",
                          "fontSize": 15,
                          "lineHeight": 20,
                        },
                      ]
                    }
                  >
                    How do I find my recovery phrase?
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                [
                  {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 16,
                  },
                ]
              }
            />
            <View
              accessibilityState={
                {
                  "disabled": true,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              hitSlop={
                {
                  "bottom": 5,
                  "left": 5,
                  "right": 5,
                  "top": 5,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FC72FF",
                  "borderColor": "transparent",
                  "borderRadius": 16,
                  "borderWidth": 1,
                  "display": "flex",
                  "opacity": 0.4,
                }
              }
            >
              <View
                sentry-label="Flex"
                style={
                  [
                    {
                      "alignItems": "stretch",
                      "display": "flex",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                      "paddingHorizontal": 12,
                      "paddingVertical": 12,
                    },
                  ]
                }
              >
                <Text
                  allowFontScaling={true}
                  maxFontSizeMultiplier={1.2}
                  style={
                    [
                      {
                        "color": "#FFFFFF",
                        "display": "flex",
                        "fontFamily": "Inter-SemiBold",
                        "fontSize": 17,
                        "lineHeight": 20,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Continue
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  ExpoLinearGradient
</View>
`;
